File: core/src/main/javascript/entityauth/EntityAuthenticationScheme.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>Entity authentication schemes.</p>
 * 
 * <p>The scheme name is used to uniquely identify entity authentication
 * schemes.</p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
    "use strict";
    
    var Class = require('../util/Class.js');
    
    /** Map of names onto schemes. */
    var schemes = {};
    
    var EntityAuthenticationScheme = module.exports = Class.create({
        /**
         * Define an entity authentication scheme with the specified name and
         * cryptographic properties.
         * 
         * @param {string} name the entity authentication scheme name.
         * @param {boolean} encrypts true if the scheme encrypts message data.
         * @param {boolean} protects true if the scheme protects message integrity.
         */
        init: function init(name, encrypts, protects) {
            // The properties.
            var props = {
                name: { value: name, writable: false, configurable: false },
                encrypts: { value: encrypts, writable: false, configurable: false },
                protectsIntegrity: { value: protects, writable: false, configurable: false },
            };
            Object.defineProperties(this, props);
            
            // Add this scheme to the map.
            schemes[name] = this;
        },
        
        /** @inheritDoc */
        toString: function toString() {
            return this.name;
        },
    });

    Class.mixin(EntityAuthenticationScheme,
    /** @lends {EntityAuthenticationScheme} */
    ({
        /** Pre-shared keys. */
        PSK : new EntityAuthenticationScheme("PSK", true, true),
        /** Pre-shared keys with entity profiles. */
        PSK_PROFILE : new EntityAuthenticationScheme("PSK_PROFILE", true, true),
        /** X.509 public/private key pair. */
        X509 : new EntityAuthenticationScheme("X509", false, true),
        /** RSA public/private key pair. */
        RSA : new EntityAuthenticationScheme("RSA", false, true),
        /** ECC public/private key pair. */
        ECC: new EntityAuthenticationScheme("ECC", false, true),
        /** Unauthenticated. */
        NONE : new EntityAuthenticationScheme("NONE", false, false),
        /** Unauthenticated suffixed. */
        NONE_SUFFIXED : new EntityAuthenticationScheme("NONE_SUFFIXED", false, false),
        /** Master token protected. */
        MT_PROTECTED : new EntityAuthenticationScheme("MT_PROTECTED", false, false),
        /** Provisioned. */
        PROVISIONED : new EntityAuthenticationScheme("PROVISIONED", false, false),
        /** ESN Migration. */
        MIGRATION : new EntityAuthenticationScheme("MIGRATION", false, false),
    }));

    /**
     * @param {string} name the entity authentication scheme name.
     * @return {?EntityAuthenticationScheme} the scheme identified by the specified name or {@code null} if
     *         there is none.
     */
    var EntityAuthenticationScheme$getScheme = function EntityAuthenticationScheme$getScheme(name) {
        return (schemes[name]) ? schemes[name] : null;
    };
    
    // Exports.
    Object.defineProperties(EntityAuthenticationScheme, {
        getScheme: { value: EntityAuthenticationScheme$getScheme, writable: false, enumerable: false, configurable: false },
    });
    Object.freeze(EntityAuthenticationScheme);
})(require, (typeof module !== 'undefined') ? module : mkmodule('EntityAuthenticationScheme'));

File: core/src/main/cpp/entityauth/EntityAuthenticationScheme.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <entityauth/EntityAuthenticationScheme.h>
#include <util/Mutex.h>

using namespace std;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace entityauth {

EntityAuthenticationScheme EntityAuthenticationScheme::PSK("PSK", true, true);
EntityAuthenticationScheme EntityAuthenticationScheme::PSK_PROFILE("PSK_PROFILE", true, true);
EntityAuthenticationScheme EntityAuthenticationScheme::X509("X509", false, true);
EntityAuthenticationScheme EntityAuthenticationScheme::RSA("RSA", false, true);
EntityAuthenticationScheme EntityAuthenticationScheme::ECC("ECC", false, true);
EntityAuthenticationScheme EntityAuthenticationScheme::NONE("NONE", false, false);
EntityAuthenticationScheme EntityAuthenticationScheme::NONE_SUFFIXED("NONE_SUFFIXED", false, false);
EntityAuthenticationScheme EntityAuthenticationScheme::MT_PROTECTED("MT_PROTECTED", false, false);
EntityAuthenticationScheme EntityAuthenticationScheme::PROVISIONED("PROVISIONED", false, false);
EntityAuthenticationScheme EntityAuthenticationScheme::INVALID("INVALID", false, false);
    
// static
StaticMslMutex& EntityAuthenticationScheme::mutex()
{
    static StaticMslMutex mutex;
    return mutex;
}
    
// static
std::map<std::string, EntityAuthenticationScheme>& EntityAuthenticationScheme::schemes()
{
    static std::map<std::string, EntityAuthenticationScheme> schemes;
    return schemes;
}

EntityAuthenticationScheme::EntityAuthenticationScheme(const string& name,
        bool encrypts, bool protects)
    : name_(name)
    , encrypts_(encrypts)
    , protects_(protects)
{
    LockGuard lockGuard(mutex());
    schemes().insert(make_pair(name, *this));
}

//static
EntityAuthenticationScheme EntityAuthenticationScheme::getScheme(const string& name)
{
    LockGuard lockGuard(mutex());
    const MapType::const_iterator it = schemes().find(name);
    return (it == schemes().end()) ? INVALID : it->second;
}

//static
vector<EntityAuthenticationScheme> EntityAuthenticationScheme::values()
{
    LockGuard lockGuard(mutex());
    vector<EntityAuthenticationScheme> v;
    for(MapType::const_iterator it = schemes().begin(); it != schemes().end(); ++it)
        if (it->second != INVALID) v.push_back(it->second);
    return v;
}

bool operator==(const EntityAuthenticationScheme& a, const EntityAuthenticationScheme& b) {
    return a.name() == b.name();
}

bool operator!=(const EntityAuthenticationScheme& a, const EntityAuthenticationScheme& b) {
    return !(a == b);
}

bool operator<(const EntityAuthenticationScheme& a, const EntityAuthenticationScheme& b) {
    return a.name() > b.name();
}

}}} // namespace netflix::msl::entityauth
